src/storage-client/src/storage_collections.rs

```diff
@@ -1824,27 +1859,167 @@ where
+            if !matches!(&existing.description.data_source, DataSource::Table { .. }) {
+                return Err(StorageError::IdentifierInvalid(existing_collection));
+            }
 
-        collection.collection_metadata.relation_desc = new_desc.clone();
-        collection.description.desc = new_desc.clone();
+            (
+                existing.collection_metadata.data_shard,
+                existing.write_frontier.clone(),
+                existing.read_capabilities.clone(),
+            )
+        };
 
-        debug!("altered table {table_id}'s RelationDesc");
+        let persist_client = self
+            .persist
+            .open(self.persist_location.clone())
+            .await
+            .unwrap();
+
+        // Evolve the schema of this shard.
+        let diagnostics = Diagnostics {
+            shard_name: existing_collection.to_string(),
+            handle_purpose: "alter_table_desc".to_string(),
+        };
+        // We map the Adapter's RelationVersion 1:1 with SchemaId.
+        let expected_schema = expected_version.into();
+        let schema_result = persist_client
+            .compare_and_evolve_schema::<SourceData, (), T, Diff>(
+                data_shard,
+                expected_schema,
+                &new_desc,
+                &UnitSchema,
+                diagnostics,
+            )
+            .await
+            .map_err(|e| StorageError::InvalidUsage(e.to_string()))?;
+        tracing::info!(
+            ?existing_collection,
+            ?new_collection,
+            ?new_desc,
+            "evolved schema"
+        );
+
+        match schema_result {
+            CaESchema::Ok(id) => id,
+            // TODO(alter_table): Better handling of these errors.
+            CaESchema::ExpectedMismatch { .. } => {
+                return Err(StorageError::Generic(anyhow::anyhow!(
+                    "schema expected mismatch, {existing_collection:?}",
+                )))
+            }
+            CaESchema::Incompatible => {
+                return Err(StorageError::Generic(anyhow::anyhow!(
+                    "schema incompatible, {existing_collection:?}"
+                )))
+            }
+        };
+
+        // Once the new schema is registered we can open new data handles.
+        let (write_handle, since_handle) = self
+            .open_data_handles(
+                &new_collection,
+                data_shard,
+                None,
+                new_desc.clone(),
+                &persist_client,
+            )
+            .await;
+
+        // TODO(alter_table): Do we need to advance the since of the table to match the time this
+        // new version was registered with txn-wal?
+
+        // Because this is a Table, we know it's managed by txn_wal, and thus it's logical write
+        // frontier is possibly in advance of the write_handle's upper. So we fast forward the
+        // write frontier to match that of the existing collection.
+        let write_frontier =
```

Comment:
The comment seems to imply that `write_handle.upper() <= existing_writer_frontier` is an invariant. Should we unconditionally use the `existing_write_frontier` then? Is there any case where that inequality doesn't hold?

Reply:
I think the only case this inequality doesn't hold is if a shard that is registered with `txn-wal` is written to outside of `txn-wal`, so misuse. I added a soft_assert here incase there is a separate edge case, e.g. on initialization or something.